designs/RestoreAndBuildCompatibility.md

```diff
@@ -0,0 +1,224 @@
+
+In recent years, NuGet restore has added many capabilities to support .NET Core. From FrameworkReference in 5.3 to support for .NET 5 TFMs in 5.8.
+In *most* situations, using an older version of NuGet.exe *will not* cause any problems at build time. 5.8 and .NET SDK 5.0.100 in particular _do_ cause problems.
+[NuGet/Home#9756](https://github.com/NuGet/Home/issues/9756) introduces a change that was needed to support .NET 5 tfms and that change in itself required a significant change to how the build tasks interpret the `project.assets.json`.
+An example is the following Developer Community [issue](https://developercommunity2.visualstudio.com/t/NuGet-Restore-build-issues---projectass/1190427).
+
+#### Newer NuGet tooling, older build tooling
+
+Say a customer has configured to get latest NuGet in their build tasks. While NuGet.exe and VisualStudio do ship on the same day, the rollout to all build machines is not atomic. The customers could end up in this mismatched situation without an explicit action.
+
+A newer NuGet version *might* generate an assets file that has more/different details than what the build tasks expect.
+A recent example is the following Developer Community [issue](https://developercommunity2.visualstudio.com/t/vsbuild-msbuild-task-failing-since-2020-11-10/1249432) which is not 100% understood yet.
+
+In this design, we are trying to improve the general mismatched tooling experience, but also consider alternatives for the particular issues customers are hitting with VS 16.8/NuGet 5.8/ SDK 5.0.100.
+
+## Who are the customers
+
+PackageReference customers
+
+## Goals
+
+* Improve the experience for mismatched versions of restore/build tooling.
+
+## Non-Goals
+
+* Introducing explicit backwards/forward compatibility. That's a lateral effort.
+
+## Solution Overview
+
+In the 2 particular problems described above the experience when things go wrong is inconsistent. It's not *always* apparent why something goes wrong.
+This proposal covers improving that warning/error experience.
+
+### NuGet output versioning
+
+When PackageReference restore is run, NuGet writes out the 3 files mentioned above. 2 of those files have a `version`.
+
+* The `project.assets.json` has a version that has historically been used for schema changes, not tooling compatibility checks. For example, aliases addition involved adding more information to the assets file, but not a schema change so this number was not incremented.
+* The `nuget.g.props` contains a `NuGetToolVersion` property.
+
+### Build Tasks warning (SDK)
+
+Given that the implied compatibility exists, the proposal is for the SDK build tasks raise a suppressible warning when the output is not in the expected shape.
+
+* Warn when the restore output is from an older tool than anticipated.
+* Warn when the restore output is from a newer tool than anticipated.
+
+There are 2 technical options and each might have a different effect on the frequency in which this warning is raised.

```

Comment:
```suggestion
There are 3 technical options and each might have a different effect on the frequency in which this warning is raised.
```

Reply:
Also, I feel like this document should pick one before being merged. Is that the intention?

Reply:
Yes, my plan is to get opinions/ideas today, and decide by Monday latest.